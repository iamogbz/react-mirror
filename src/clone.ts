import { compare } from "specificity";
import {
    camelToSpinalCase,
    getPseudoElementSelector,
    splitSelectors,
} from "./utils";

export type CloneOptions = {
    styles?: React.CSSProperties;
    class?: string;
};

export function copyStyles(
    sourceElt: HTMLElement,
    targetElt: HTMLElement,
    options?: CloneOptions,
): void {
    const styleDecls: { [key: string]: CSSStyleDeclaration } = {};
    for (let i = 0; i < document.styleSheets.length; i++) {
        const styleSheet = document.styleSheets[i] as CSSStyleSheet;
        const rules = styleSheet.rules || styleSheet.cssRules;
        for (let j = 0; j < rules.length; j++) {
            const rule = rules[j] as CSSStyleRule;
            // test each selector in a group separately
            // accounts for bug with specificity library where
            // `.classA, .classB` fails when compared with `.classB`
            // https://github.com/keeganstreet/specificity/issues/4
            for (const selectorText of splitSelectors(rule.selectorText)) {
                // also match pseudo selectors
                const selector = selectorText.replace(
                    getPseudoElementSelector(selectorText),
                    "",
                );
                try {
                    if (sourceElt?.matches(selector || "*")) {
                        styleDecls[selectorText] = rule.style;
                    }
                } catch (e) {
                    // eslint-disable-next-line no-console
                    console.error(e.message, { selectorText });
                }
            }
        }
    }
    // ensures only the cloned styles are applied to element
    const singleClassName = targetElt.className.replace(" ", "-");
    targetElt.className = `${singleClassName}${options?.class || ""}`;
    // style element used for transfering pseudo styles
    const styleElt = document.createElement("style");
    styleElt.type = "text/css";
    for (const selector of Object.keys(styleDecls).sort(compare)) {
        const styleDecl = styleDecls[selector];
        // store pseudo element styles in style element
        const pseudoElementSelector = getPseudoElementSelector(selector);
        if (pseudoElementSelector) {
            styleElt.innerHTML += `.${targetElt.className}${pseudoElementSelector} { ${styleDecl.cssText} }\n`;
        } else {
            for (let i = 0; i < styleDecl.length; i++) {
                const propName = styleDecl[i];
                const propValue = styleDecl.getPropertyValue(propName);
                targetElt.style.setProperty(propName, propValue);
            }
        }
    }
    // append style element used for transfering pseudo styles
    styleElt.innerHTML && targetElt.appendChild(styleElt);
}

export function cloneNodeWithStyles(node: Node, options: CloneOptions): Node {
    const clone = node.cloneNode(false);
    // check that node is html element before cloning styles
    const cloneElt = clone as HTMLElement;
    const nodeElt = node as HTMLElement;
    if (nodeElt.style && cloneElt.style) {
        copyStyles(nodeElt, cloneElt, options);
        for (const [prop, value] of Object.entries(options.styles || {})) {
            cloneElt.style.setProperty(camelToSpinalCase(prop), value);
        }
    }
    return clone;
}

export function deepCloneWithStyles(node: Node, options: CloneOptions): Node {
    const clone = cloneNodeWithStyles(node, options);
    for (const child of node.childNodes) {
        clone.appendChild(deepCloneWithStyles(child, options));
    }
    return clone;
}
