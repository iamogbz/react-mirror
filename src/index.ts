export { Frame, FrameProps } from "./components/Frame";
export { Mirror, MirrorProps } from "./components/Mirror";
export { Reflection, ReflectionProps } from "./components/Reflection";
export { useMirror } from "./hooks/useMirror";
