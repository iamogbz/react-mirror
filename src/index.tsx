import * as React from "react";
import * as ReactDOM from "react-dom";

export function Mirror({
    reflect,
}: {
    reflect: React.ReactInstance | null;
}): JSX.Element {
    // ref to element in which reflection will be framed
    const [frame, ref] = React.useState<HTMLDivElement | null>(null);
    // real dom node of react element being reflected
    // eslint-disable-next-line react/no-find-dom-node
    const real = React.useMemo(() => ReactDOM.findDOMNode(reflect), [reflect]);
    // update the reflection to match the real node
    const update = React.useCallback(() => {
        if (!frame || !real) return;
        frame.innerHTML = "";
        frame.appendChild(real.cloneNode(true));
    }, [frame, real]);
    // start of the reflection
    React.useEffect(update, [update]);
    // mutation observer single instance
    const observer = React.useMemo(() => new MutationObserver(update), [
        update,
    ]);
    // observe or disconnect when real node is set or unmounted
    React.useEffect(() => {
        if (!real || !observer) return undefined;
        observer.observe(real, {
            attributes: true,
            childList: true,
            subtree: true,
            characterData: true,
        });
        return (): void => observer.disconnect();
    }, [real, observer]);
    // return frame element
    return <div ref={ref} style={{ pointerEvents: "none" }} />;
}

export function useMirror(): [
    React.Dispatch<unknown>,
    JSX.Element | undefined,
] {
    const [node, ref] = React.useState(null);
    return [ref, <Mirror key="mirror" reflect={node} />];
}
