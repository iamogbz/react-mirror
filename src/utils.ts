export function randomString(length: number): string {
    let str = "";
    while (str.length < length) {
        str += Math.random()
            .toString(36)
            .substr(2, length - str.length);
    }
    return str;
}

export function camelToSpinalCase(camelCase: string): string {
    return camelCase.replace(/([A-Z])/g, "-$1").toLocaleLowerCase();
}

export function splitSelectors(selectors: string): string[] {
    if (!selectors) {
        return [];
    }
    if (isAtRule(selectors)) {
        return [selectors];
    }
    const splitted = [];
    let quotes = 0;
    let parens = 0;
    let angulars = 0;
    let soFar = "";
    for (let i = 0, len = selectors.length; i < len; i++) {
        const char = selectors[i];
        if (char === "(") {
            parens += 1;
        } else if (char === ")") {
            parens -= 1;
        } else if (char === "[") {
            angulars += 1;
        } else if (char === "]") {
            angulars -= 1;
        } else if (char === '"') {
            quotes += 1;
        } else if (char === ",") {
            if (!parens && !angulars && !(quotes % 2)) {
                splitted.push(soFar.trim());
                soFar = "";
                continue;
            }
        }
        soFar += char;
    }
    splitted.push(soFar.trim());
    return splitted.filter(Boolean);
}

function isAtRule(selector: string): boolean {
    return selector.indexOf("@") === 0;
}

export function getPseudoElementSelector(selector: string): string {
    const pseudoElementSelector = /::\w+(\(\w+\))?/g;
    return selector.match(pseudoElementSelector)?.[0] ?? "";
}
