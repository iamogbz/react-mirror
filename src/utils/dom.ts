export function getChildren(element?: Element | Text) {
    const children: (Element | Text)[] = [];
    if (isElement(element)) {
        element.childNodes.forEach((childNode) => {
            if (isElement(childNode) || isText(childNode)) {
                children.push(childNode);
            }
        });
    }
    return children;
}

export function getAttributes(element?: Element | Text) {
    const attributes: Record<string, string> = {};
    if (isElement(element)) {
        Array.from(element.attributes).forEach((attr) => {
            attributes[attr.name] = attr.value;
        });
    }
    attributes.value = getValue(element);
    return attributes;
}

function getValue(node?: Node) {
    return (node as HTMLInputElement)?.value;
}

export function getBounds(node?: Node) {
    if (isElement(node)) {
        return node.getBoundingClientRect();
    }
    if (isText(node)) {
        const range = document.createRange?.();
        range?.selectNodeContents(node);
        return range?.getBoundingClientRect() ?? new DOMRect();
    }
    return new DOMRect();
}

export function isElement(node?: Node): node is Element {
    return node ? !isText(node) : false;
}

export function isText(node?: Node): node is Text {
    return node?.nodeType === Node.TEXT_NODE;
}
