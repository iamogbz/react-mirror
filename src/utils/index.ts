export function randomString(length: number) {
    let str = "";
    while (str.length < length) {
        str += Math.random()
            .toString(36)
            .substr(2, length - str.length);
    }
    return str;
}

export function spinalToCamelCase(spinalCase: string): string {
    return spinalCase.replace(/-[a-z]/g, (m) => m.substring(1).toUpperCase());
}
