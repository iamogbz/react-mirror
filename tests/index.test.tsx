import * as React from "react";
// import { act } from "react-test-renderer";
import { prettyDOM } from "@testing-library/dom";
import { render, cleanup } from "@testing-library/react";
import { renderHook, act } from "@testing-library/react-hooks";
import { Mirror, useMirror } from "../src";

describe("Hook", (): void => {
    it("returns an empty reflection", (): void => {
        const { result } = renderHook(() => useMirror());
        const [, reflection] = result.current;
        expect(reflection).toMatchSnapshot();
    });

    it("starts reflecting node", (): void => {
        const { result } = renderHook(() => useMirror());
        const [ref] = result.current;
        act(() => void ref(document.createElement("div")));
        const [, reflection] = result.current;
        expect(reflection).toMatchSnapshot();
    });
});

describe("Component", (): void => {
    beforeAll(() => {
        jest.spyOn(Math, "random").mockReturnValue(0.123456789);
    });
    afterEach(cleanup);
    afterAll(jest.restoreAllMocks);

    it("frames an empty reflection", (): void => {
        const result = render(
            <Mirror className="test-mirror-frame" reflect={null} />,
        );
        expect(result.baseElement).toMatchSnapshot();
    });

    it("renders reflection with styles", () => {
        /** add document style */
        const domStyle = document.createElement("style");
        document.head.appendChild(domStyle);
        domStyle.innerHTML = `
            body {
                font-family: "san-serif";
                font-size: 1.2em;
            }
            .classOne {
                height: 10px;
            }
            .classTwo {
                font-size: 1.3em;
                display: block;
                width: 40px;
                margin: 0 auto;
            }
            .classThree::after {
                content: '';
                background: red;
                width: 5px;
                height: 5px;
            }
            .classOne .classTwo {
                width: 20px;
            }
        `;
        /** add nodes that will be mirrored mirror */
        const domNode = document.createElement("div");
        document.body.appendChild(domNode);
        const node1 = document.createElement("div");
        domNode.appendChild(node1);
        node1.className = "classOne";
        const node2 = document.createElement("span");
        node1.appendChild(node2);
        node2.className = "classTwo";
        const node3 = document.createElement("p");
        domNode.appendChild(node3);
        node3.innerHTML = "Mock text node";
        node3.className = "classThree";
        /** mirror nodes and check results */
        const result = render(<Mirror reflect={domNode} />);
        expect(result.baseElement).toMatchSnapshot();
        /** clean up */
        domStyle.remove();
        domNode.remove();
    });
});
