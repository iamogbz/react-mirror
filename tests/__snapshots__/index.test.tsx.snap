// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component frames an empty reflection 1`] = `
<body>
  <div>
    <div
      class="test-mirror-frame"
    />
  </div>
</body>
`;

exports[`Component renders reflection with styles 1`] = `
<body>
  <div>
    <div
      attr="[value"
      class="class1 one"
    >
      <span
        class="class2 two"
      />
    </div>
    <p
      class="class3 three"
    >
      Mock text node
    </p>
  </div>
  <div>
    <div
      class="mirrorFrame_4fzzzxj"
    >
      <div
        class="_4fzzzxj"
        style="pointer-events: none;"
      >
        <style
          type="text/css"
        >
          ._4fzzzxj::before { position: absolute; }
._4fzzzxj::after { content: ''; }

        </style>
        <div
          attr="[value"
          class="class1-one_4fzzzxj"
          style="height: 10px; pointer-events: none;"
        >
          <style
            type="text/css"
          >
            .class1-one_4fzzzxj::before { position: absolute; }
.class1-one_4fzzzxj::after { content: ''; }

          </style>
          <span
            class="class2-two_4fzzzxj"
            style="font-size: 1.3em; display: block; width: 20px; margin: 0px auto; pointer-events: none;"
          >
            <style
              type="text/css"
            >
              .class2-two_4fzzzxj::before { position: absolute; }
.class2-two_4fzzzxj::after { content: ''; }

            </style>
          </span>
        </div>
        <p
          class="class3-three_4fzzzxj"
          style="pointer-events: none;"
        >
          <style
            type="text/css"
          >
            .class3-three_4fzzzxj::before { position: absolute; }
.class3-three_4fzzzxj::after { content: ''; }
.class3-three_4fzzzxj::after { background: red; width: 5px; height: 5px; }

          </style>
          Mock text node
        </p>
      </div>
      <style
        type="text/css"
      >
        .mirrorFrame_4fzzzxj::before { position: absolute; }
.mirrorFrame_4fzzzxj::after { content: ''; }
.mirrorFrame_4fzzzxj::before { content: 'mock text'; }

      </style>
    </div>
  </div>
</body>
`;

exports[`Hook returns an empty reflection 1`] = `<Mirror />`;

exports[`Hook starts reflecting node 1`] = `
<Mirror
  className="test-mirror"
  reflect={<div />}
/>
`;
